### PR TITLE
[`typing`] Add `find_assigned_value` helper func to `typing.rs` to retrieve value of a given variable `id`

### PR DESCRIPTION
**Pull request by qdegraaf**
## Summary

Adds `find_assigned_value` a function which gets the `&Expr` assigned to a given `id` if one exists in the semantic model. 

Open TODOs:

- [ ] Handle `binding.kind.is_unpacked_assignment()`: I am bit confused by this one. The snippet from its documentation does not appear to be counted as an unpacked assignment and the only ones I could find for which that was true were invalid Python like:
```python
x, y = 1 
```
- [ ] How to handle AugAssign. Can we combine statements like:
```python
(a, b) = [(1, 2, 3), (4,)]
a += (6, 7)
```
to get the full value for a? Code currently just returns `None` for these assign types

- [ ] Multi target assigns
```python
m_c = (m_d, m_e) = (0, 0)
trio.sleep(m_c)  # OK
trio.sleep(m_d)  # TRIO115
trio.sleep(m_e)  # TRIO115
```

## Test Plan

Used the function in two rules:

- `TRIO115`
- `PERF101`

Expanded both their fixtures for explicit multi target check

### CODE REVIEW

**Comment by qdegraaf** [comment]
@charliermarsh I'm thinking the `binding.kind.is_unpacked_assignment()` can be left as a TODO for later when it is necessary? Let me know if that is acceptable. I think it would be good to review current set-up and logic first regardless, before expanding further.

File: crates/ruff_python_semantic/src/analyze/typing.rs [hunk 1]
```diff
@@ -568,3 +568,99 @@ pub fn resolve_assignment<'a>(
...
+    let binding = semantic.binding(binding_id);
+    if binding.kind.is_assignment() || binding.kind.is_named_expr_assignment() {
+        let parent_id = binding.source?;
+        let parent = semantic.statement(parent_id);
+        match parent {
+            Stmt::Assign(ast::StmtAssign { value, targets, .. }) => match value.as_ref() {
+                Expr::Tuple(ast::ExprTuple { elts, .. })
+                | Expr::List(ast::ExprList { elts, .. })
+                | Expr::Set(ast::ExprSet { elts, .. }) => {
+                    if let Some(target) = targets.iter().next() {
```
Comment (charliermarsh) [inline crates/ruff_python_semantic/src/analyze/typing.rs:594]: What happens here in the case of multi-target assignments?

Comment (qdegraaf) [reply crates/ruff_python_semantic/src/analyze/typing.rs:594]: Good catch. It works for 

```python
x = y = 0
```

Added example of that to TRIO115 fixture

Does not work for 

```python
(x, y) = [a, b] = (0, 0)
```

Which will only manage to retrieve the values for `x` and `y` right now. 

Would it be an idea to set up unit tests for this and the other new func in the module? If so, is there a leading example of non plugin unit tests I can have a look at?

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/analyze/typing.rs:594]: Is it able to detect `y` in `x = y = 0`?

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/analyze/typing.rs:594]: Yeah, we don't have great infrastructure for this. I'd probably just recommend adding more trio test cases as a means of testing this behavior.

**Comment by charliermarsh** [comment]
I think there's one more case to consider:

```python
m_c = (m_d, m_e) = (0, 0)
trio.sleep(m_c)  # OK
trio.sleep(m_d)  # TRIO115
trio.sleep(m_e)  # TRIO115
```

Right now, these _all_ get matched to `(0, 0)`.

File: crates/ruff_python_semantic/src/analyze/typing.rs (hunk 1 above)
Comment (qdegraaf) [reply crates/ruff_python_semantic/src/analyze/typing.rs:592]: > Is it able to detect `y` in `x = y = 0`?

Yeah this one works. But if one of the targets is a collection there are issues. e.g.

```python
a = (x, y) = (0, 0)
(a, b) = [c, d] = [0, 0]
```
Not entirely sure how to best fix that. Flag it as a multi assign if the top-level targets has a length larger than 1 and add a separate branch for that? Or are there single target scenarios that could be mistaken then?

**Comment by qdegraaf** [comment]
> I think there's one more case to consider:
> 
> ```python
> m_c = (m_d, m_e) = (0, 0)
> trio.sleep(m_c)  # OK
> trio.sleep(m_d)  # TRIO115
> trio.sleep(m_e)  # TRIO115
> ```
> 
> Right now, these _all_ get matched to `(0, 0)`.

Ho missed this one when I posted https://github.com/astral-sh/ruff/pull/8583#discussion_r1402583307 

Added it to TODOs. Gonna look at this again, see if there's a way to be aware of the multi target nature of the assign before drilling down.

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**